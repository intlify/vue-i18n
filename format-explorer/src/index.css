body {
  --bg: #1D1F21;
  --fg: #fff;
  --border: #333;
  --in-bg: white;
  --in-fg: black;
  --in-border: #666;
  margin: 0;
  background-color: var(--bg);
}

#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: var(--fg);
  margin: 0;
  width: 100vw;
  height: 100vh;
  overflow: hidden;
}
