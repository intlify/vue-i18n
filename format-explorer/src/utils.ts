import { customRef } from 'vue'
import type { Ref } from 'vue'

export function useDebouncedRef<T>(value: T, delay = 200): Ref<T> {
  let timeout: ReturnType<typeof setTimeout>
  return customRef<T>((track, trigger) => {
    return {
      get() {
        track()
        return value
      },
      set(newValue: T) {
        clearTimeout(timeout)
        timeout = setTimeout(() => {
          value = newValue
          trigger()
        }, delay)
      }
    }
  })
}
