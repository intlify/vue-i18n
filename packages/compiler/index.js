'use strict'

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./dist/compiler.cjs.prod.js')
} else {
  module.exports = require('./dist/compiler.cjs.js')
}
