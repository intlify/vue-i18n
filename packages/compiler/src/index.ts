import { isObject } from '@intlify/shared'

export type Foo = 'Foo'

console.log('hello compiler!', isObject({}))
