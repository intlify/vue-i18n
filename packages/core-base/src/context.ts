import {
  warn,
  isString,
  isArray,
  isBoolean,
  isRegExp,
  isFunction,
  isPlainObject,
  isObject
} from '@intlify/shared'
import { DevToolsTimelineEvents } from './debugger/constants'
import { CoreWarnCodes, getWarnMessage } from './warnings'

import type { Path } from '@intlify/message-resolver'
import type { CompileOptions } from '@intlify/message-compiler'
import type {
  LinkedModifiers,
  PluralizationRules,
  MessageProcessor,
  MessageFunction,
  MessageType
} from '@intlify/runtime'
import type { DevToolsEmitter } from './debugger/constants'
import type {
  NumberFormat,
  DateTimeFormat,
  DateTimeFormats as DateTimeFormatsType,
  NumberFormats as NumberFormatsType
} from './types'

/** @VueI18nGeneral */
export type Locale = string

/** @VueI18nGeneral */
export type FallbackLocale =
  | Locale
  | Locale[]
  | { [locale in string]: Locale[] }
  | false

/** @VueI18nGeneral */
export type LocaleMessageValue<Message = string> =
  | string
  | MessageFunction<Message>
  | LocaleMessageDictionary<Message>
  | LocaleMessageArray<Message>

/** @VueI18nGeneral */
export type LocaleMessageDictionary<Message = string> = {
  [property: string]: LocaleMessageValue<Message>
}
/** @VueI18nGeneral */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface LocaleMessageArray<Message = string>
  extends Array<LocaleMessageValue<Message>> {}
/** @VueI18nGeneral */
export type LocaleMessages<Message = string> = Record<
  Locale,
  LocaleMessageDictionary<Message>
>

export type CoreMissingType = 'translate' | 'datetime format' | 'number format'

export type CoreMissingHandler<Message = string> = (
  context: CoreCommonContext<Message>,
  locale: Locale,
  key: Path,
  type: CoreMissingType,
  ...values: unknown[]
) => string | void
/** @VueI18nGeneral */
export type PostTranslationHandler<Message = string> = (
  translated: MessageType<Message>
) => MessageType<Message>

export type MessageCompiler<Message = string> = (
  source: string,
  options?: CompileOptions
) => MessageFunction<Message>

export interface CoreOptions<Message = string> {
  locale?: Locale
  fallbackLocale?: FallbackLocale
  messages?: LocaleMessages<Message>
  datetimeFormats?: DateTimeFormatsType
  numberFormats?: NumberFormatsType
  modifiers?: LinkedModifiers<Message>
  pluralRules?: PluralizationRules
  missing?: CoreMissingHandler<Message>
  missingWarn?: boolean | RegExp
  fallbackWarn?: boolean | RegExp
  fallbackFormat?: boolean
  unresolving?: boolean
  postTranslation?: PostTranslationHandler<Message>
  processor?: MessageProcessor<Message>
  warnHtmlMessage?: boolean
  escapeParameter?: boolean
  messageCompiler?: MessageCompiler<Message>
  onWarn?: (msg: string, err?: Error) => void
}

export interface CoreInternalOptions {
  __datetimeFormatters?: Map<string, Intl.DateTimeFormat>
  __numberFormatters?: Map<string, Intl.NumberFormat>
  __emitter?: DevToolsEmitter // for vue-devtools timeline event
}

export interface CoreCommonContext<Message = string> {
  locale: Locale
  fallbackLocale: FallbackLocale
  missing: CoreMissingHandler<Message> | null
  missingWarn: boolean | RegExp
  fallbackWarn: boolean | RegExp
  fallbackFormat: boolean
  unresolving: boolean
  onWarn(msg: string, err?: Error): void
}

export interface CoreTranslationContext<Messages = {}, Message = string>
  extends CoreCommonContext<Message> {
  messages: Messages
  modifiers: LinkedModifiers<Message>
  pluralRules?: PluralizationRules
  postTranslation: PostTranslationHandler<Message> | null
  processor: MessageProcessor<Message> | null
  warnHtmlMessage: boolean
  escapeParameter: boolean
  messageCompiler: MessageCompiler<Message> | null
}

export interface CoreDateTimeContext<DateTimeFormats = {}, Message = string>
  extends CoreCommonContext<Message> {
  datetimeFormats: DateTimeFormats
}

export interface CoreNumberContext<NumberFormats = {}, Message = string>
  extends CoreCommonContext<Message> {
  numberFormats: NumberFormats
}

export interface CoreContext<
  Messages = {},
  DateTimeFormats = {},
  NumberFormats = {},
  Message = string
> extends CoreTranslationContext<Messages, Message>,
    CoreDateTimeContext<DateTimeFormats, Message>,
    CoreNumberContext<NumberFormats, Message> {}

export interface CoreInternalContext {
  __datetimeFormatters: Map<string, Intl.DateTimeFormat>
  __numberFormatters: Map<string, Intl.NumberFormat>
  __localeChainCache?: Map<Locale, Locale[]>
  __emitter?: DevToolsEmitter // for vue-devtools timeline event
}

export const NOT_REOSLVED = -1

export const MISSING_RESOLVE_VALUE = ''

function getDefaultLinkedModifiers<
  Message = string
>(): LinkedModifiers<Message> {
  return {
    upper: (val: Message): MessageType<Message> =>
      (isString(val) ? val.toUpperCase() : val) as MessageType<Message>,
    lower: (val: Message): MessageType<Message> =>
      (isString(val) ? val.toLowerCase() : val) as MessageType<Message>,
    // prettier-ignore
    capitalize: (val: Message): MessageType<Message> =>
      (isString(val)
        ? `${val.charAt(0).toLocaleUpperCase()}${val.substr(1)}`
        : val) as MessageType<Message>
  }
}

let _compiler: unknown | null

export function registerMessageCompiler<Message>(
  compiler: MessageCompiler<Message>
): void {
  _compiler = compiler
}

export function createCoreContext<
  Message = string,
  Options extends CoreOptions<Message> = object,
  Messages extends Record<
    keyof Options['messages'],
    LocaleMessageDictionary<Message>
  > = Record<keyof Options['messages'], LocaleMessageDictionary<Message>>,
  DateTimeFormats extends Record<
    keyof Options['datetimeFormats'],
    DateTimeFormat
  > = Record<keyof Options['datetimeFormats'], DateTimeFormat>,
  NumberFormats extends Record<
    keyof Options['numberFormats'],
    NumberFormat
  > = Record<keyof Options['numberFormats'], NumberFormat>
>(
  options: Options = {} as Options
): CoreContext<
  Options['messages'],
  Options['datetimeFormats'],
  Options['numberFormats'],
  Message
> {
  // setup options
  const locale = isString(options.locale) ? options.locale : 'en-US'
  const fallbackLocale =
    isArray(options.fallbackLocale) ||
    isPlainObject(options.fallbackLocale) ||
    isString(options.fallbackLocale) ||
    options.fallbackLocale === false
      ? options.fallbackLocale
      : locale
  const messages = isPlainObject(options.messages)
    ? options.messages
    : ({ [locale]: {} } as Messages)
  const datetimeFormats = isPlainObject(options.datetimeFormats)
    ? options.datetimeFormats
    : ({ [locale]: {} } as DateTimeFormats)
  const numberFormats = isPlainObject(options.numberFormats)
    ? options.numberFormats
    : ({ [locale]: {} } as NumberFormats)
  const modifiers = Object.assign(
    {} as LinkedModifiers<Message>,
    options.modifiers || ({} as LinkedModifiers<Message>),
    getDefaultLinkedModifiers<Message>()
  )
  const pluralRules = options.pluralRules || {}
  const missing = isFunction(options.missing) ? options.missing : null
  const missingWarn =
    isBoolean(options.missingWarn) || isRegExp(options.missingWarn)
      ? options.missingWarn
      : true
  const fallbackWarn =
    isBoolean(options.fallbackWarn) || isRegExp(options.fallbackWarn)
      ? options.fallbackWarn
      : true
  const fallbackFormat = !!options.fallbackFormat
  const unresolving = !!options.unresolving
  const postTranslation = isFunction(options.postTranslation)
    ? options.postTranslation
    : null
  const processor = isPlainObject(options.processor) ? options.processor : null
  const warnHtmlMessage = isBoolean(options.warnHtmlMessage)
    ? options.warnHtmlMessage
    : true
  const escapeParameter = !!options.escapeParameter
  const messageCompiler = isFunction(options.messageCompiler)
    ? options.messageCompiler
    : _compiler
  const onWarn = isFunction(options.onWarn) ? options.onWarn : warn

  // setup internal options
  const internalOptions = options as CoreInternalOptions
  const __datetimeFormatters = isObject(internalOptions.__datetimeFormatters)
    ? internalOptions.__datetimeFormatters
    : new Map<string, Intl.DateTimeFormat>()
  const __numberFormatters = isObject(internalOptions.__numberFormatters)
    ? internalOptions.__numberFormatters
    : new Map<string, Intl.NumberFormat>()

  const context = {
    locale,
    fallbackLocale,
    messages,
    datetimeFormats,
    numberFormats,
    modifiers,
    pluralRules,
    missing,
    missingWarn,
    fallbackWarn,
    fallbackFormat,
    unresolving,
    postTranslation,
    processor,
    warnHtmlMessage,
    escapeParameter,
    messageCompiler,
    onWarn,
    __datetimeFormatters,
    __numberFormatters
  } as CoreContext<
    Options['messages'],
    Options['datetimeFormats'],
    Options['numberFormats'],
    Message
  >

  // for vue-devtools timeline event
  if (__DEV__) {
    ;((context as unknown) as CoreInternalContext).__emitter =
      internalOptions.__emitter != null ? internalOptions.__emitter : undefined
  }

  return context
}

/** @internal */
export function isTranslateFallbackWarn(
  fallback: boolean | RegExp,
  key: Path
): boolean {
  return fallback instanceof RegExp ? fallback.test(key) : fallback
}

/** @internal */
export function isTranslateMissingWarn(
  missing: boolean | RegExp,
  key: Path
): boolean {
  return missing instanceof RegExp ? missing.test(key) : missing
}

/** @internal */
export function handleMissing<Message = string>(
  context: CoreCommonContext<Message>,
  key: Path,
  locale: Locale,
  missingWarn: boolean | RegExp,
  type: CoreMissingType
): unknown {
  const { missing, onWarn } = context

  // for vue-devtools timeline event
  if (__DEV__) {
    const emitter = ((context as unknown) as CoreInternalContext).__emitter
    if (emitter) {
      emitter.emit(DevToolsTimelineEvents.MISSING, {
        locale,
        key,
        type,
        groupId: `${type}:${key}`
      })
    }
  }

  if (missing !== null) {
    const ret = missing(context, locale, key, type)
    return isString(ret) ? ret : key
  } else {
    if (__DEV__ && isTranslateMissingWarn(missingWarn, key)) {
      onWarn(getWarnMessage(CoreWarnCodes.NOT_FOUND_KEY, { key, locale }))
    }
    return key
  }
}

/** @internal */
export function getLocaleChain<Message = string>(
  ctx: CoreCommonContext<Message>,
  fallback: FallbackLocale,
  start: Locale
): Locale[] {
  const context = (ctx as unknown) as CoreInternalContext

  if (!context.__localeChainCache) {
    context.__localeChainCache = new Map()
  }

  let chain = context.__localeChainCache.get(start)
  if (!chain) {
    chain = []

    // first block defined by start
    let block: unknown = [start]

    // while any intervening block found
    while (isArray(block)) {
      block = appendBlockToChain(chain, block, fallback)
    }

    // prettier-ignore
    // last block defined by default
    const defaults = isArray(fallback)
      ? fallback
      : isPlainObject(fallback)
        ? fallback['default']
          ? fallback['default']
          : null
        : fallback

    // convert defaults to array
    block = isString(defaults) ? [defaults] : defaults
    if (isArray(block)) {
      appendBlockToChain(chain, block, false)
    }
    context.__localeChainCache.set(start, chain)
  }

  return chain
}

function appendBlockToChain(
  chain: Locale[],
  block: Locale[],
  blocks: FallbackLocale
): unknown {
  let follow: unknown = true
  for (let i = 0; i < block.length && isBoolean(follow); i++) {
    const locale = block[i]
    if (isString(locale)) {
      follow = appendLocaleToChain(chain, block[i], blocks)
    }
  }
  return follow
}

function appendLocaleToChain(
  chain: Locale[],
  locale: Locale,
  blocks: FallbackLocale
): unknown {
  let follow: unknown
  const tokens = locale.split('-')
  do {
    const target = tokens.join('-')
    follow = appendItemToChain(chain, target, blocks)
    tokens.splice(-1, 1)
  } while (tokens.length && follow === true)
  return follow
}

function appendItemToChain(
  chain: Locale[],
  target: Locale,
  blocks: FallbackLocale
): unknown {
  let follow: unknown = false
  if (!chain.includes(target)) {
    follow = true
    if (target) {
      follow = target[target.length - 1] !== '!'
      const locale = target.replace(/!/g, '')
      chain.push(locale)
      if (
        (isArray(blocks) || isPlainObject(blocks)) &&
        (blocks as any)[locale] // eslint-disable-line @typescript-eslint/no-explicit-any
      ) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        follow = (blocks as any)[locale]
      }
    }
  }
  return follow
}

/** @internal */
export function updateFallbackLocale<Message = string>(
  ctx: CoreCommonContext<Message>,
  locale: Locale,
  fallback: FallbackLocale
): void {
  const context = (ctx as unknown) as CoreInternalContext
  context.__localeChainCache = new Map()
  getLocaleChain(ctx, fallback, locale)
}
