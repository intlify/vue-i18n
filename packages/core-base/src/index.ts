import { getGlobalThis } from '@intlify/shared'

export {
  CompileError,
  CompileErrorCodes,
  createCompileError
} from '@intlify/message-compiler'
export * from './resolver'
export * from './runtime'
export * from './context'
export * from './fallbacker'
export * from './compile'
export * from './translate'
export * from './datetime'
export * from './number'
export { getWarnMessage, CoreWarnCodes } from './warnings'
export { CoreError, CoreErrorCodes, createCoreError } from './errors'
export * from './types'
export * from './devtools'

if (__ESM_BUNDLER__ && !__TEST__) {
  if (typeof __FEATURE_PROD_INTLIFY_DEVTOOLS__ !== 'boolean') {
    getGlobalThis().__INTLIFY_PROD_DEVTOOLS__ = false
  }
}
