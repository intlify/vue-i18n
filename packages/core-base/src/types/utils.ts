/* eslint-disable @typescript-eslint/no-explicit-any */

import type { LocaleMessage } from '../context'
import type { DateTimeFormat, NumberFormat } from './intl'

export type IsUnion<T, B = T> = T extends B
  ? [B] extends [T]
    ? false
    : true
  : never

// prettier-ignore
export type UnionToIntersection<U> = (U extends any ? (arg: U) => void : never) extends (arg: infer I) => void
  ? I
  : never

export type LastInUnion<U> = UnionToIntersection<
  U extends unknown ? (x: U) => 0 : never
> extends (x: infer L) => 0
  ? L
  : never

export type UnionToTuple<U, Last = LastInUnion<U>> = [U] extends [never]
  ? []
  : [...UnionToTuple<Exclude<U, Last>>, Last]

export type LocaleRecord<T extends any[], R> = {
  [K in T[number]]: R
}

export type First<T extends readonly any[]> = T[0]

type __ResourcePath<T, Key extends keyof T> = Key extends string
  ? T[Key] extends Record<string, any>
    ?
        | `${Key}.${__ResourcePath<T[Key], Exclude<keyof T[Key], keyof any[]>> &
            string}`
        | `${Key}.${Exclude<keyof T[Key], keyof any[]> & string}`
    : never
  : never
type _ResourcePath<T> = __ResourcePath<T, keyof T> | keyof T
export type ResourcePath<T> = _ResourcePath<T> extends string | keyof T
  ? _ResourcePath<T>
  : keyof T

export type ResourceValue<
  T,
  P extends ResourcePath<T>
> = P extends `${infer Key}.${infer Rest}`
  ? Key extends keyof T
    ? Rest extends ResourcePath<T[Key]>
      ? ResourceValue<T[Key], Rest>
      : never
    : never
  : P extends keyof T
  ? T[P]
  : never

export type PickupLocales<
  T extends Record<string, any>,
  K = keyof T
> = K extends string ? K : never

export type PickupKeys<
  T extends Record<string, any>,
  K = keyof T
> = K extends string ? ResourcePath<T[K]> : never

type __ResourceFormatPath<T, Key extends keyof T> = Key extends string
  ? T[Key] extends Record<string, any>
    ?  | `${Key}`
    : never
  : never
type _ResourceFormatPath<T> = __ResourceFormatPath<T, keyof T> | keyof T
export type ResourceFormatPath<T> = _ResourceFormatPath<T> extends string | keyof T
  ? _ResourceFormatPath<T>
  : keyof T

export type PickupFormatKeys<
    T extends Record<string, any>,
    K = keyof T
> = K extends string ? ResourceFormatPath<T[K]> : never

export type PickupFallbackLocales<T extends any[]> = T[number] | `${T[number]}!`

export type FallbackLocales<Locales = 'en-US'> =
  | Locales
  | Array<Locales>
  | {
      [locale in string]: Array<PickupFallbackLocales<UnionToTuple<Locales>>>
    }
  | false

// prettier-ignore
type LocaleParamsType<T, R> = T extends IsUnion<T>
  ? T
  : T extends string
    ? T
    : R

// prettier-ignore
export type SchemaParams<T, Message = string> = T extends readonly any[]
  ? { message: First<T>, datetime: DateTimeFormat, number: NumberFormat }
  : T extends { message?: infer M, datetime?: infer D, number?: infer N }
    ? {
      message: M extends LocaleMessage<Message> ? M : LocaleMessage<Message>,
      datetime: D extends DateTimeFormat ? D : DateTimeFormat,
      number: N extends NumberFormat ? N : NumberFormat
    }
    : {
      message: LocaleMessage<Message>,
      datetime: DateTimeFormat,
      number: NumberFormat
    }

// prettier-ignore
export type LocaleParams<T, Default = 'en-US'> = T extends IsUnion<T>
  ? { messages: T, datetimeFormats: T, numberFormats: T }
  : T extends { messages?: infer M, datetimeFormats?: infer D, numberFormats?: infer N }
    ? {
      messages: LocaleParamsType<M, Default>,
      datetimeFormats: LocaleParamsType<D, Default>,
      numberFormats: LocaleParamsType<N, Default>
    }
    : T extends string
      ? { messages: T, datetimeFormats: T, numberFormats: T }
      : { messages: Default, datetimeFormats: Default, numberFormats: Default }

/* eslint-enable @typescript-eslint/no-explicit-any */