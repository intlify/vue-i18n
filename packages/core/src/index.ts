import {
  registerMessageCompiler,
  compileToFunction,
  registerMessageResolver,
  resolveValue,
  registerLocaleFallbacker,
  fallbackWithLocaleChain
} from '@intlify/core-base'

// register message compiler at @intlify/core
registerMessageCompiler(compileToFunction)

// register message resolver at @intlify/core
registerMessageResolver(resolveValue)

// register fallback locale at @intlify/core
registerLocaleFallbacker(fallbackWithLocaleChain)

export * from '@intlify/core-base'
