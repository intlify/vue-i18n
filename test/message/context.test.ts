import { compile } from '../../src/message/compiler'
import { createMessageContext } from '../../src/message/context'
import { isString } from '../../src/utils'

describe('text', () => {
  test('basic', () => {
    const msg = compile('hello world')
    const ctx = createMessageContext()
    expect(msg(ctx)).toMatch('hello world')
  })

  test('multline', () => {
    const msg = compile('hello\n world')
    const ctx = createMessageContext()
    expect(JSON.stringify(msg(ctx))).toMatch(`hello\\n world`)
  })
})

describe('list', () => {
  test('basic', () => {
    const msg = compile('hi {0} !')
    const ctx = createMessageContext({
      list: ['kazupon']
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('multiple', () => {
    const msg = compile('{0} {1} !')
    const ctx = createMessageContext({
      list: ['hi', 'kazupon']
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('nothing interpolation data', () => {
    const msg = compile('hi {0} !')
    const ctx = createMessageContext()
    expect(msg(ctx)).toMatch(`hi  !`)
  })
})

describe('named', () => {
  test('basic', () => {
    const msg = compile('hi {name} !')
    const ctx = createMessageContext({
      named: { name: 'kazupon' }
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('multiple', () => {
    const msg = compile('{greeting} {name} !')
    const ctx = createMessageContext({
      named: {
        greeting: 'hi',
        name: 'kazupon'
      }
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('nothing interpolation data', () => {
    const msg = compile('hi {name} !')
    const ctx = createMessageContext()
    expect(msg(ctx)).toMatch(`hi  !`)
  })
})

describe('linked', () => {
  test('key', () => {
    const msg = compile('hi @:name !')
    const ctx = createMessageContext({
      messages: {
        name: ctx => 'kazupon' // eslint-disable-line
      }
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('list', () => {
    const msg = compile('hi @:{0} !')
    const ctx = createMessageContext({
      list: ['name'],
      messages: {
        name: ctx => 'kazupon' // eslint-disable-line
      }
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('named', () => {
    const msg = compile('hi @:{name} !')
    const ctx = createMessageContext({
      named: {
        name: 'nickname'
      },
      messages: {
        nickname: ctx => 'kazupon' // eslint-disable-line
      }
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('modifier', () => {
    const msg = compile(`hi @.upper:{'name'} !`)
    const ctx = createMessageContext({
      modifiers: {
        upper: (str: string): string => str.toUpperCase()
      },
      messages: {
        name: ctx => 'kazupon' // eslint-disable-line
      }
    })
    expect(msg(ctx)).toMatch(`hi KAZUPON !`)
  })

  test('no modifier', () => {
    const msg = compile(`hi @.upper:{'name'} !`)
    const ctx = createMessageContext({
      messages: {
        name: ctx => 'kazupon' // eslint-disable-line
      }
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('resolve from function', () => {
    const msg = compile('hi @:name !')
    const ctx = createMessageContext({
      messages: str => ctx => 'kazupon' // eslint-disable-line
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('reoslve from parent', () => {
    const parentMsg = compile('kazupon')
    const parent = createMessageContext({
      messages: {
        name: parentMsg
      }
    })
    const msg = compile('hi @:name !')
    const ctx = createMessageContext({
      parent
    })
    expect(msg(ctx)).toMatch(`hi kazupon !`)
  })

  test('not found message', () => {
    const msg = compile('hi @:{0} !')
    const ctx = createMessageContext({
      list: ['name'],
      messages: {}
    })
    expect(msg(ctx)).toMatch(`hi  !`)
  })
})

describe('plural', () => {
  test('basic', () => {
    const msg = compile('no apples | one apple  |  too much apples  ')
    const ctx0 = createMessageContext({ pluralIndex: 0 })
    const ctx1 = createMessageContext({ pluralIndex: 1 })
    const ctx2 = createMessageContext({ pluralIndex: 2 })
    expect(msg(ctx0)).toMatch(`no apples`)
    expect(msg(ctx1)).toMatch(`one apple`)
    expect(msg(ctx2)).toMatch(`too much apples`)
  })

  test('pre-defined', () => {
    const msg = compile('no apples | one apple | {count} apples')
    const ctxPreDefined = createMessageContext({ pluralIndex: 10 })
    const ctxNamed = createMessageContext({
      pluralIndex: 10,
      named: {
        count: 20
      }
    })
    expect(msg(ctxPreDefined)).toMatch(`10 apples`)
    expect(msg(ctxNamed)).toMatch(`20 apples`)
  })

  test('complex', () => {
    const msg = compile('no apples | {n} @.lower:{unit} | {n}　apples')
    const ctx = createMessageContext({
      modifiers: {
        lower: (str: string): string => str.toLowerCase()
      },
      named: {
        unit: 'apple',
        n: 1
      },
      messages: {
        'apple': ctx => 'APPLE' // eslint-disable-line
      },
      pluralIndex: 1
    })
    expect(msg(ctx)).toMatch(`1 apple`)
  })

  test('custom choice rule', () => {
    const pluralRules = {
      ru: (choice, choicesLength) => {
        if (choice === 0) {
          return 0
        }

        const teen = choice > 10 && choice < 20
        const endsWithOne = choice % 10 === 1
        if (!teen && endsWithOne) {
          return 1
        }
        if (!teen && choice % 10 >= 2 && choice % 10 <= 4) {
          return 2
        }

        return choicesLength < 4 ? 2 : 3
      }
    }
    const msg = compile('0 машин | {n} машина | {n} машины | {n} машин')
    const ctx1 = createMessageContext({
      locale: 'ru',
      pluralIndex: 1,
      pluralRules
    })
    const ctx2 = createMessageContext({
      locale: 'ru',
      pluralIndex: 2,
      pluralRules
    })
    const ctx3 = createMessageContext({
      locale: 'ru',
      pluralIndex: 4,
      pluralRules
    })
    const ctx4 = createMessageContext({
      locale: 'ru',
      pluralIndex: 12,
      pluralRules
    })
    const ctx5 = createMessageContext({
      locale: 'ru',
      pluralIndex: 21,
      pluralRules
    })
    const ctx6 = createMessageContext({
      locale: 'ja',
      pluralIndex: 21,
      pluralRules
    })
    expect(msg(ctx1)).toMatch(`1 машина`)
    expect(msg(ctx2)).toMatch(`2 машины`)
    expect(msg(ctx3)).toMatch(`4 машины`)
    expect(msg(ctx4)).toMatch(`12 машин`)
    expect(msg(ctx5)).toMatch(`21 машина`)
    expect(msg(ctx6)).toMatch(`21 машин`)
  })
})

describe('custom process', () => {
  const createVNode = (text: string) => ({ text })
  const normalize = (values: unknown[]): unknown =>
    values.map(val => (isString(val) ? createVNode(val) : val))
  const interpolate = (val: unknown): unknown => val

  test('simple text', () => {
    const msg = compile('hello')
    const ctx = createMessageContext({
      processor: {
        normalize,
        interpolate
      }
    })
    expect(msg(ctx)).toMatchSnapshot()
  })

  test('list', () => {
    const msg = compile('hi, {0}!')
    const ctx = createMessageContext({
      list: [createVNode('kazupon')],
      processor: {
        normalize,
        interpolate
      }
    })
    expect(msg(ctx)).toMatchSnapshot()
  })

  test('named', () => {
    const msg = compile('hi {name} !')
    const ctx = createMessageContext({
      named: { name: createVNode('kazupon') },
      processor: {
        normalize,
        interpolate
      }
    })
    expect(msg(ctx)).toMatchSnapshot()
  })

  test('linked', () => {
    const msg = compile(`hi @.upper:{'name'} !`)
    const ctx = createMessageContext({
      processor: {
        normalize,
        interpolate,
        type: 'vnode'
      },
      modifiers: {
        upper: (val: unknown, type: string): unknown =>
          `${type}:${(val as { text: string }).text.toUpperCase()}`
      },
      messages: {
        name: ctx => createVNode('kazupon') // eslint-disable-line
      }
    })
    expect(msg(ctx)).toMatchSnapshot()
  })

  test('plural', () => {
    const msg = compile('no apples | {n} @.lower:{unit} | {n}　apples')
    const ctx = createMessageContext({
      processor: {
        normalize,
        interpolate,
        type: 'vnode'
      },
      modifiers: {
        lower: (val: unknown, type: string): unknown =>
          `${type}:${(val as { text: string }).text.toLowerCase()}`
      },
      named: {
        unit: 'apple',
        n: '1'
      },
      messages: {
        'apple': ctx => createVNode('APPLE') // eslint-disable-line
      },
      pluralIndex: 1
    })
    expect(msg(ctx)).toMatchSnapshot()
  })
})

describe('edge cases', () => {
  test(`text: '&nbsp;'`, () => {
    const msg = compile('&nbsp;')
    const ctx = createMessageContext()
    expect(msg(ctx)).toMatch(`&nbsp;`)
  })

  test(`plural: ' | foo | '`, () => {
    const msg = compile(` | foo | `, {
      onError(error) {
        expect({ ...error, message: error.message }).toMatchSnapshot('error')
      }
    })
    const ctx = createMessageContext()
    expect(msg(ctx)).toMatch('')
  })
})
