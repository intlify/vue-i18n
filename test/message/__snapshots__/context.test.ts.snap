// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom process linked 1`] = `
Array [
  Object {
    "text": "hi ",
  },
  Object {
    "text": "vnode:KAZUPON",
  },
  Object {
    "text": " !",
  },
]
`;

exports[`custom process list 1`] = `
Array [
  Object {
    "text": "hi, ",
  },
  Object {
    "text": "kazupon",
  },
  Object {
    "text": "!",
  },
]
`;

exports[`custom process named 1`] = `
Array [
  Object {
    "text": "hi ",
  },
  Object {
    "text": "kazupon",
  },
  Object {
    "text": " !",
  },
]
`;

exports[`custom process plural 1`] = `
Array [
  Object {
    "text": "1",
  },
  Object {
    "text": " ",
  },
  Object {
    "text": "vnode:apple",
  },
]
`;

exports[`custom process simple text 1`] = `
Array [
  Object {
    "text": "hello",
  },
]
`;

exports[`edge cases plural: ' | foo | ': error 1`] = `
Object {
  "code": 10,
  "domain": "parser",
  "location": Object {
    "end": Object {
      "column": 10,
      "line": 1,
      "offset": 9,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "message": "Plural must have messages",
}
`;
